Pretty printing errors

mtzguido opened the pull request:
This PR makes Pulse use the new error doc framework in F*, and improves the two most common errors with it: 'cannot prove vprop' and 'unmatched vprop in post'. See this example from `L0Crypto.fst`

Trying to free something twice:
<img width="509" alt="Screenshot 2023-09-18 173905" src="https://github.com/FStarLang/steel/assets/4195583/60c85016-b042-4da6-9404-b73a92c0f231">

Forgetting to free something:
<img width="481" alt="Screenshot 2023-09-18 174707" src="https://github.com/FStarLang/steel/assets/4195583/18a89421-6e55-42d4-a65f-33fb460cebfa">

nikswamy commented:
This is a big improvement, thanks!

While we're here, can I nit pick at a few things in the error message?

* Can we remove the string "Tactic logged issue"?
* Can we drop the word "vprops" from the messages? 
* Do we need the backticks around the assertions? e.g., A.pts_to adigest (reveal s), instead of enclosing within \` \`
* Can we rename "post hint" to, say, "annotated postcondition"? "post hint" is internal, implementation lingo.
* Can the last message say:

- Error in proving postcondition
- Inferred postcondition additionally contains:
A.pts_to ...

mtzguido commented:
> This is a big improvement, thanks!
> 
> While we're here, can I nit pick at a few things in the error message?
> 
> * Can we remove the string "Tactic logged issue"?

It's easy to remove, and I see it looks very weird here, but I added in F* to make sure that errors raised by a tactic can be clearly distinguished from normal errors. I'm having second thoughts about it now though, it's not like tactics/plugins are really adversarial, so I maybe it's fine to drop it (and maybe --debug forces it to print or something like that).

> * Can we drop the word "vprops" from the messages?

Done

> * Do we need the backticks around the assertions? e.g., A.pts_to adigest (reveal s), instead of enclosing within ` `

Removed. I nested them 2 extra spaces deep though. WDYT?

> * Can we rename "post hint" to, say, "annotated postcondition"? "post hint" is internal, implementation lingo.
> * Can the last message say:
>   
>   * Error in proving postcondition
>   * Inferred postcondition additionally contains:
>     A.pts_to ...

Edited. Some captures here:

![2023-09-18 20:55:35_6](https://github.com/FStarLang/steel/assets/4195583/fed51079-020b-432c-beae-e7a766d9c82f)
![2023-09-18 20:55:53_7](https://github.com/FStarLang/steel/assets/4195583/dab3d168-4295-40f3-9ac2-047813374a18)
![2023-09-18 20:56:05_8](https://github.com/FStarLang/steel/assets/4195583/6e0407d9-39db-455f-852e-a1c73087cd72)

Other minor ideas:
- The second one could say "Did you forget to free?"
- For the last one, I'm not sure the initial context+goal are so useful, at least nowadays

nikswamy commented:
> The second one could say "Did you forget to free?"

This looks good to me too

mtzguido commented:
Updated! The initial context+goal will print if `--ext pulse:initial_solver_state` is "true" (set to something other than `""`, `"false"` or `"0"`). The `env_on_err` option is still available too.

I'll add the debug level toggle on a separate PR.
